File: app/src/main/java/me/ykrank/s1next/view/dialog/QuotePostPageParserDialogFragment.java
package me.ykrank.s1next.view.dialog;

import android.os.Bundle;

import com.google.common.base.Optional;
import com.google.common.base.Preconditions;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.model.ThreadLink;
import me.ykrank.s1next.util.ErrorUtil;
import me.ykrank.s1next.view.activity.PostListActivity;
import me.ykrank.s1next.widget.track.event.PageEndEvent;
import me.ykrank.s1next.widget.track.event.PageStartEvent;
import rx.Observable;

/**
 * A {@link ProgressDialogFragment} parses post post page for thread.
 */
public final class QuotePostPageParserDialogFragment extends ProgressDialogFragment<String> {

    public static final String TAG = QuotePostPageParserDialogFragment.class.getName();

    private static final String ARG_THREAD_LINK = "thread_link";

    private boolean mShouldFinishActivity = true;

    public static QuotePostPageParserDialogFragment newInstance(ThreadLink threadLink) {
        QuotePostPageParserDialogFragment fragment = new QuotePostPageParserDialogFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ARG_THREAD_LINK, threadLink);
        fragment.setArguments(bundle);

        return fragment;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if (mShouldFinishActivity) {
            getActivity().finish();
        }
    }

    @Override
    protected Observable<String> getSourceObservable() {
        ThreadLink threadLink = Preconditions.checkNotNull(getArguments().getParcelable(
                ARG_THREAD_LINK));
        return mS1Service.getQuotePostResponseBody(threadLink.getThreadId(),
                threadLink.getQuotePostId().get()).map(voidResponse ->
                voidResponse.raw().request().url().toString());
    }

    @Override
    protected void onNext(String url) {
        Optional<Integer> jumpPage = parseQuotePostPage(url);
        if (jumpPage.isPresent()) {
            ThreadLink threadLink = Preconditions.checkNotNull(getArguments().getParcelable(
                    ARG_THREAD_LINK));
            ThreadLink threadLinkWithJumpPage = new ThreadLink.Builder(threadLink.getThreadId())
                    .jumpPage(jumpPage.get())
                    .quotePostId(threadLink.getQuotePostId().get())
                    .build();
            PostListActivity.startPostListActivity(getActivity(), threadLinkWithJumpPage);
        } else {
            ThreadLinkInvalidPromptDialogFragment.newInstance(getContext(),
                    R.string.dialog_message_quote_not_found).show(getFragmentManager(),
                    ThreadLinkInvalidPromptDialogFragment.TAG);
            mShouldFinishActivity = false;
        }
    }

    @Override
    protected void onError(Throwable throwable) {
        ThreadLinkInvalidPromptDialogFragment.newInstance(getContext(), ErrorUtil.parse(throwable))
                .show(getFragmentManager(), ThreadLinkInvalidPromptDialogFragment.TAG);
        mShouldFinishActivity = false;
    }

    @Override
    protected CharSequence getProgressMessage() {
        return getString(R.string.dialog_message_processing);
    }

    /**
     * Parses redirect link in order to get post post page.
     *
     * @param url The redirect link.
     */
    private Optional<Integer> parseQuotePostPage(String url) {
        // example: http://bbs.saraba1st.com/2b/forum.php?mod=viewthread&tid=1074030&page=1#pid27217893
        Pattern pattern = Pattern.compile("page=(\\d+)");
        Matcher matcher = pattern.matcher(url);
        if (matcher.find()) {
            return Optional.of(Integer.parseInt(matcher.group(1)));
        }
        return Optional.absent();
    }

    @Override
    public void onResume() {
        super.onResume();
        trackAgent.post(new PageStartEvent("弹窗-链接解析进度条-" + TAG));
    }

    @Override
    public void onPause() {
        trackAgent.post(new PageEndEvent("弹窗-链接解析进度条-" + TAG));
        super.onPause();
    }
}

File: app/src/main/java/me/ykrank/s1next/view/dialog/QuotePostPageParserDialogFragment.kt
package me.ykrank.s1next.view.dialog

import android.os.Build
import android.os.Bundle
import androidx.lifecycle.lifecycleScope
import com.github.ykrank.androidtools.extension.toast
import io.reactivex.Single
import kotlinx.coroutines.launch
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.model.link.ThreadLink
import me.ykrank.s1next.util.ErrorUtil
import me.ykrank.s1next.view.page.post.postlist.PostListActivity

/**
 * A [ProgressDialogFragment] parses post post page for thread.
 */
class QuotePostPageParserDialogFragment : ProgressDialogFragment<String>() {

    private var threadLink: ThreadLink? = null
    private var mShouldFinishActivity = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            threadLink = requireArguments().getParcelable(ARG_THREAD_LINK, ThreadLink::class.java)
        } else {
            threadLink = requireArguments().getParcelable(ARG_THREAD_LINK)
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        if (mShouldFinishActivity) {
            activity?.finish()
        }
    }

    override fun getSourceObservable(): Single<String> {
        val threadLink = this.threadLink ?: return Single.error(IllegalStateException())
        return mS1Service.getQuotePostResponseBody(threadLink.threadId,
            threadLink.quotePostId
        ).map { voidResponse -> voidResponse.raw().request.url.toString() }
    }

    override fun onNext(data: String) {
        val threadLink = this.threadLink ?: return
        val jumpPage = parseQuotePostPage(data)
        if (jumpPage != null) {
            val threadLinkWithJumpPage = ThreadLink.Builder(threadLink.threadId)
                    .jumpPage(jumpPage)
                .quotePostId(threadLink.quotePostId)
                    .build()
            activity?.let {
                PostListActivity.start(it, threadLinkWithJumpPage)
            }
        } else {
            ThreadLinkInvalidPromptDialogFragment.newInstance(context,
                getString(R.string.dialog_message_quote_not_found)
            ).show(childFragmentManager, ThreadLinkInvalidPromptDialogFragment.TAG)
            mShouldFinishActivity = false
        }
    }

    override fun onError(throwable: Throwable) {
        val context = context ?: return
        lifecycleScope.launch {
            val errorMsg = ErrorUtil.parse(context, throwable)
            if (isVisible) {
                ThreadLinkInvalidPromptDialogFragment.newInstance(context, errorMsg)
                    .show(childFragmentManager, ThreadLinkInvalidPromptDialogFragment.TAG)
                mShouldFinishActivity = false
            } else {
                App.get().toast(errorMsg)
            }
        }
    }

    override fun getProgressMessage(): CharSequence? {
        return getString(R.string.dialog_message_processing)
    }

    /**
     * Parses redirect link in order to get post post page.

     * @param url The redirect link.
     */
    private fun parseQuotePostPage(url: String): Int? {
        // example: http://bbs.saraba1st.com/2b/forum.php?mod=viewthread&tid=1074030&page=1#pid27217893
        val page = "page=(\\d+)".toRegex().find(url)?.groupValues?.getOrNull(1)
        return page?.toIntOrNull()
    }

    companion object {

        val TAG = QuotePostPageParserDialogFragment::class.java.simpleName

        private const val ARG_THREAD_LINK = "thread_link"

        fun newInstance(threadLink: ThreadLink): QuotePostPageParserDialogFragment {
            val fragment = QuotePostPageParserDialogFragment()
            val bundle = Bundle()
            bundle.putParcelable(ARG_THREAD_LINK, threadLink)
            fragment.arguments = bundle

            return fragment
        }
    }
}
